File: day_6/part_1.rb
def check_four(sub_string, full_string)
  uniq_string = sub_string[0, 4].split("").uniq
  puts "Unique values #{uniq_string}"
  if uniq_string.length < 4 
    check_four(sub_string[1..], full_string)
  else
    index = full_string.index(sub_string)
    puts puts "Number of characters processed: #{index + 4}"
  end
end

text = File.read("input.txt")
check_four(text, text)

File: day_6/part_1.exs
defmodule Day6.Lib do
  def check_four([a, b, c, d | _] = char_list) do
    if Enum.uniq_by([a, b, c, d], fn {letter, _} -> letter end) |> length() < 4 do
      [_ | rest_of_list] = char_list
      check_four(rest_of_list)
    else
      {_, last_index} = d
      IO.puts("Number of characters processed: #{last_index + 1}")
    end
  end
end

with file <- File.read!("input.txt") do
  String.graphemes(file)
  |> Enum.with_index()
  |> Day6.Lib.check_four()
end
